src/python/dxpy/bindings/dxfile_functions.py

```diff
@@ -466,3 +466,101 @@ def upload_string(to_upload, media_type=None, keep_open=False, wait_on_close=Fal
         handler.close(block=wait_on_close, **remaining_kwargs)
 
     return handler
+
+def static_var(var_name, initial_value):
+    def decorate(func):
+        setattr(func, var_name, initial_value)
+        return func
+    return decorate
+
+@static_var("folders_cache", {})
+def list_subfolders(project, path, usecache=True, recurse=True):
+    '''
+    :param project: Project ID to use as context for the listing
+    :type project: string
+    :param path: Subtree root path
+    :type path: string
+    :param usecache: Use folders listing cache
+    :type usecache: boolean
+    :param recurse: Return a complete subfolders tree
+    :type recurse: boolean
+
+    Returns a list of subfolders for the remote *path* (included to the result) of the *project*.
+
+    Example::
+
+        list_subfolders("project-xxxx", folder="/input", usecache=False)
+
+    '''
+    if usecache:
+        if project not in list_subfolders.folders_cache:
+            list_subfolders.folders_cache[project] = dxpy.get_handler(project).describe(input_params={'folders': True})['folders']
+        project_folders = list_subfolders.folders_cache[project]
+    else:
+        project_folders = dxpy.get_handler(project).describe(input_params={'folders': True})['folders']
+    # TODO: support shell-style path globbing (i.e. /a*/c matches /ab/c but not /a/b/c)
+    # return pathmatch.filter(project_folders, os.path.join(path, '*'))
+    if recurse:
+        return (f for f in project_folders if f.startswith(path))
+    else:
+        return (f for f in project_folders if f.startswith(path) and '/' not in f[len(path)+1:])
+
+def download_folder(project, destdir, folder="/", overwrite=False, chunksize=dxfile.DEFAULT_BUFFER_SIZE,
+        usecache=True, **kwargs):
```

Comment:
One potential problem with this scheme for caching is that if you have a long-running process, data in the cache could grow arbitrarily stale. Previously the cache was at least scoped to a single call to "download", but this is no longer the case. Perhaps the caching logic could be removed entirely, unless there is some obvious place where there are going to be many repeated calls to the same subfolder in the course of performing some common operation.